// react-async-select/build/index.es.js
import React, { useRef, useEffect, useState } from 'react';
import styled, { css } from 'styled-components';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */

function __makeTemplateObject(cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
}

var Maybe = function (_a) {
    var isOpen = _a.isOpen, children = _a.children;
    return isOpen ? children : null;
};

var ListContainer = styled.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  ", "\n"], ["\n  ",
    "\n"])), function (_a) {
    var listContainerStyle = _a.listContainerStyle;
    return css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n      ", "\n    "], ["\n      ", "\n    "])), listContainerStyle);
});
var Item = styled.div(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  ", "\n"], ["\n  ",
    "\n"])), function (_a) {
    var selected = _a.selected, listItemStyle = _a.listItemStyle, listItemSelectStyle = _a.listItemSelectStyle;
    return selected
        ? css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n          ", "\n          ", "\n        "], ["\n          ", "\n          ", "\n        "])), listItemStyle, listItemSelectStyle) : css(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n          ", "\n        "], ["\n          ", "\n        "])), listItemStyle);
});
var EmptyData = styled.div(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  padding: 0.5rem 0;\n  color: gray;\n  text-align: center;\n"], ["\n  padding: 0.5rem 0;\n  color: gray;\n  text-align: center;\n"])));
var List = function (_a) {
    var message = _a.message, listIndex = _a.listIndex, debouncedList = _a.debouncedList, setListIndex = _a.setListIndex, handleSelectItem = _a.handleSelectItem, handleBlur = _a.handleBlur, listContainerStyle = _a.listContainerStyle, listItemStyle = _a.listItemStyle, listItemSelectStyle = _a.listItemSelectStyle;
    var listRef = useRef(null);
    var handleClickItem = function (e) {
        var index = Number(e.currentTarget.dataset.index);
        handleSelectItem(debouncedList[index]);
        setListIndex(0);
        handleBlur();
    };
    var handleMouseEnterListItme = function (e) {
        var index = Number(e.currentTarget.dataset.index);
        setListIndex(index);
    };
    useEffect(function () {
        var list = listRef.current;
        var children = list.children;
        if (list && children.length > 1) {
            var child = children[listIndex];
            var scroll_1 = child.offsetTop;
            list.scrollTop = scroll_1 - 100;
        }
    }, [listIndex, listRef]);
    return (React.createElement(ListContainer, { ref: listRef, listContainerStyle: listContainerStyle }, debouncedList.length ? (debouncedList.map(function (listItem, index) { return (React.createElement(Item, { onMouseEnter: handleMouseEnterListItme, key: index, "data-index": index, onClick: handleClickItem, selected: index === listIndex, listItemStyle: listItemStyle, listItemSelectStyle: listItemSelectStyle }, listItem.name)); })) : (React.createElement(EmptyData, null, message || "empty"))));
};
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;

var containerDefaultStyle = {
    width: "10rem",
    height: "2rem",
    padding: "0 0.5rem",
    border: "1px solid black",
};
var inputDefaultStyle = {
    width: "100%",
    height: "100%",
    border: "none",
    outline: "none",
    "box-sizing": "border-box",
};
var listContainerDefaultStyle = {
    "overflow-y": "auto",
    position: "absolute",
    top: "2rem",
    width: "100%",
    "max-height": "25rem",
    border: "1px solid black",
    "background-color": "white",
    "box-sizing": "border-box",
    "z-index": "2",
};
var listItemDefaultStyle = {
    padding: "0.3rem 0.2rem",
};
var listItemDefaultSelectStyle = {
    color: "white",
    cursor: "pointer",
};
var AsyncContainer = styled.div(templateObject_1$1 || (templateObject_1$1 = __makeTemplateObject(["\n  position: relative;\n  background-color: white;\n  z-index: 2;\n"], ["\n  position: relative;\n  background-color: white;\n  z-index: 2;\n"])));
var AsyncWrapper = styled.div(templateObject_3$1 || (templateObject_3$1 = __makeTemplateObject(["\n  display: flex;\n  align-items: center;\n  justify-content: space-between;\n  box-sizing: border-box;\n\n  ", ";\n"], ["\n  display: flex;\n  align-items: center;\n  justify-content: space-between;\n  box-sizing: border-box;\n\n  ",
    ";\n"])), function (_a) {
    var containerStyle = _a.containerStyle;
    return css(templateObject_2$1 || (templateObject_2$1 = __makeTemplateObject(["\n      ", "\n    "], ["\n      ", "\n    "])), containerStyle);
});
var AsyncInputWrapper = styled.div(templateObject_4$1 || (templateObject_4$1 = __makeTemplateObject(["\n  position: relative;\n  width: 100%;\n  height: 100%;\n"], ["\n  position: relative;\n  width: 100%;\n  height: 100%;\n"])));
var DisplayWrapper = styled.div(templateObject_5$1 || (templateObject_5$1 = __makeTemplateObject(["\n  position: absolute;\n  top: 0;\n  left: 0;\n  display: flex;\n  align-items: center;\n  width: 100%;\n  height: 100%;\n"], ["\n  position: absolute;\n  top: 0;\n  left: 0;\n  display: flex;\n  align-items: center;\n  width: 100%;\n  height: 100%;\n"])));
var Displayedvalue = styled.div(templateObject_6$1 || (templateObject_6$1 = __makeTemplateObject(["\n  width: 100%;\n  background-color: white;\n  white-space: nowrap;\n  text-overflow: ellipsis;\n  overflow-x: hidden;\n"], ["\n  width: 100%;\n  background-color: white;\n  white-space: nowrap;\n  text-overflow: ellipsis;\n  overflow-x: hidden;\n"])));
var AsyncInput = styled.input(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  ", "\n"], ["\n  ",
    "\n"])), function (_a) {
    var inputStyle = _a.inputStyle;
    return css(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n      ", "\n    "], ["\n      ", "\n    "])), inputStyle);
});
var Async = function (_a) {
    var inputValue = _a.inputValue, displayedValue = _a.displayedValue, handleChangeInput = _a.handleChangeInput, handleSelectItem = _a.handleSelectItem, debouncedList = _a.debouncedList, message = _a.message, style = _a.style;
    var inputRef = useRef(null);
    var _b = useState(false), isOpenList = _b[0], setIsOpenList = _b[1];
    var _c = useState(false), isFocused = _c[0], setIsFocused = _c[1];
    var _d = useState(0), listIndex = _d[0], setListIndex = _d[1];
    var displayedList = inputValue ? debouncedList : [];
    var containerStyle = style && style.containerStyle ? style.containerStyle(containerDefaultStyle) : containerDefaultStyle;
    var inputStyle = style && style.inputStyle ? style.inputStyle(inputDefaultStyle) : inputDefaultStyle;
    var listContainerStyle = style && style.listContainerStyle ? style.listContainerStyle(listContainerDefaultStyle) : listContainerDefaultStyle;
    var listItemStyle = style && style.listItemStyle ? style.listItemStyle(listItemDefaultStyle) : listItemDefaultStyle;
    var listItemSelectStyle = style && style.listItemSelectStyle
        ? style.listItemSelectStyle(listItemDefaultSelectStyle)
        : listItemDefaultSelectStyle;
    var handleFocus = function () {
        if (inputRef.current) {
            inputRef.current.focus();
        }
        setIsFocused(true);
        setIsOpenList(true);
    };
    var handleBlur = function () {
        if (inputRef.current) {
            inputRef.current.blur();
        }
        setIsFocused(false);
        setIsOpenList(false);
        handleChangeInput("");
    };
    var handleChange = function (e) {
        var value = e.target.value;
        handleChangeInput(value);
    };
    var onInputBlur = function (e) {
        e.preventDefault();
        if (isOpenList) {
            setTimeout(function () {
                handleBlur();
            }, 200);
        }
    };
    var onInputClick = function () {
        if (isFocused) {
            handleBlur();
        }
        else {
            handleFocus();
        }
    };
    var handleKeyDown = function (e) {
        var key = e.key;
        var index;
        if (key === "Escape") {
            handleBlur();
            return;
        }
        if (debouncedList.length) {
            if (key === "Enter") {
                handleSelectItem(debouncedList[listIndex]);
                setListIndex(0);
                handleBlur();
                return;
            }
            if (key === "ArrowDown") {
                index = listIndex < debouncedList.length - 1 ? listIndex + 1 : 0;
                setListIndex(index);
            }
            if (key === "ArrowUp") {
                index = listIndex >= 1 ? listIndex - 1 : debouncedList.length - 1;
                setListIndex(index);
            }
        }
    };
    return (React.createElement(AsyncContainer, null,
        React.createElement(AsyncWrapper, { containerStyle: containerStyle, onClick: onInputClick },
            React.createElement(AsyncInputWrapper, null,
                displayedValue && !isFocused ? (React.createElement(DisplayWrapper, { onClick: handleFocus },
                    React.createElement(Displayedvalue, null, displayedValue))) : null,
                React.createElement(AsyncInput, { ref: inputRef, value: inputValue, placeholder: "\uAC80\uC0C9\uC5B4\uB97C \uC785\uB825\uD574\uC8FC\uC138\uC694.", onBlur: onInputBlur, onChange: handleChange, onKeyDown: handleKeyDown, inputStyle: inputStyle }))),
        React.createElement(Maybe, { isOpen: isOpenList },
            React.createElement(List, { message: message, listIndex: listIndex, debouncedList: displayedList, handleSelectItem: handleSelectItem, handleBlur: handleBlur, setListIndex: setListIndex, listContainerStyle: listContainerStyle, listItemStyle: listItemStyle, listItemSelectStyle: listItemSelectStyle }))));
};
var templateObject_1$1, templateObject_2$1, templateObject_3$1, templateObject_4$1, templateObject_5$1, templateObject_6$1, templateObject_7, templateObject_8;

export default Async;
